spring-integration-core/src/main/java/org/springframework/integration/support/SmartLifecycleRoleController.java

```diff
@@ -185,7 +190,76 @@ public void stopLifecyclesInRole(String role) {
 		}
 	}
 
-	private void addLazyLifecycles() {
+	/**
+	 * Return a collection of the roles currently managed by this controller.
+	 * @return the roles.
+	 * @since 4.3.8
+	 */
+	public Collection<String> getRoles() {
+		if (this.lazyLifecycles.size() > 0) {
+			addLazyLifecycles();
+		}
+		return new ArrayList<>(this.lifecycles.keySet());
+	}
+
+	/**
+	 * Return true if all endpoints in the role are running.
+	 * @param role the role.
+	 * @return true if at least one endpoint in the role, and all are running.
+	 * @since 4.3.8
+	 */
+	public boolean allEndpointsRunning(String role) {
+		Map<String, Boolean> status = getEndpointsRunningStatus(role);
+		if (status.isEmpty()) {
+			return false;
+		}
+		else {
+			return status.values().stream().allMatch(b -> b);
+		}
+	}
+
+	/**
+	 * Return true if none of the endpoints in the role are running or if
+	 * there are no endpoints in the role.
+	 * @param role the role.
+	 * @return true if there are no endpoints or none are running.
+	 * @since 4.3.8
+	 */
+	public boolean noEndpointsRunning(String role) {
+		Map<String, Boolean> status = getEndpointsRunningStatus(role);
+		if (status.isEmpty()) {
+			return true;
+		}
+		else {
+			return status.values().stream().allMatch(b -> !b);
+		}
+	}
+
+	/**
+	 * Return the running status of each endpoint in the role.
+	 * @param role the role.
+	 * @return A map of component names : running status
+	 * @since 4.3.8
+	 */
+	public Map<String, Boolean> getEndpointsRunningStatus(String role) {
+		if (this.lazyLifecycles.size() > 0) {
+			addLazyLifecycles();
+		}
+		if (!this.lifecycles.containsKey(role)) {
+			return Collections.emptyMap();
+		}
+		Map<String, Boolean> runningStatus = new HashMap<>();
+		AtomicInteger index = new AtomicInteger();
+		this.lifecycles.get(role)
+				.stream()
+				.forEach(e -> {
```

Comment:
OTOH, there is something like this:
````java
return this.lifecycles.get(role)
		.stream()
		.collect(Collectors.toMap(e ->
						(e instanceof NamedComponent)
								? ((NamedComponent) e).getComponentName()
								: (e.getClass().getSimpleName() + "#" + index.getAndIncrement()),
				SmartLifecycle::isRunning));
````
http://stackoverflow.com/questions/20363719/java-8-listv-into-mapk-v